docs/internal/types.md

```diff
@@ -0,0 +1,236 @@
+ * Unions: a type that can be two or more types, ex. `(A | B)`.
+ * Intersections: a type that must implement all types, ex. `(A & imm)`.
+
+Capabilities offer language and runtime guarantees:
+ * `iso`, or `isolated`, means this is a pointer to an exclusive area of memory that only this pointer has access to in the entire program.
+ * `mut`, or `mutable`, means this object can be changed at will, which means it must be inside an area of memory you have exclusive access to, ie. traced back to an `iso` you hold the rights.
+ * `imm`, or `immutable`, means no one can write to, so anyone can read from at any time. You can make `iso` regions immutable but not the other way round, as there's no guarantees who's still reading from it.
+ * `readonly` is a local capability that is used when you want to make sure an external function can't modify, but you still hold a mutable reference.
+
+Note that type capabilities, unions and intersections are types themselves.
+So `(A & imm)` is completely different from `(A & mut)` or `(A & iso)`.
+Specifically, you cannot convert `mut` to `imm` (as you would with a `const` cast in C++), or vice versa.
+You can freeze an `iso` region into `imm`, but unfreeze an `imm` region.
+Most importantly, capabilities are not type traits like C++'s `const int * const`, they have specific non-trivial runtime semantics.
+
+The type system guarantees that:
+ * Structural soundness: Interfaces and classes form valid concrete types, operations and functions take in the correct (sometimes inferred) types.
+ * Capability guarantees are followed: The compiler checks operations on types, the runtime guarantees the semantics is followed.
+ * Runtime safety: A program that compiles is a program that runs safely, without crashes, traps or unhandled exceptions, including asynchronous behaviour, but not necessarily producing the right result or doing what you expect.
+
+## Classes and Interfaces
+
+Classes and interfaces have a similar structure:
+ * Fields: elements of a specific type that have a runtime space allocated.
+ * Methods: functions that belong to a type.
+
+The declaration is trivial:
+```ts
+// Helps receivers know they can convert the object to a string representation
+interface Serialisable
+{
+  // All derived classes will have the common field
+  var common : F64;
+
+  // Derived classes must implement this method
+  toString() : String;
+};
+
+class MyType : Serialisable
+{
+  // Inherits `common`
+  let field : I32;
+
+  // Some other method
+  innerSum(x : I32) : I32
+  {
+    let res : I32 = x + field;
+    res
+  }
+
+  // Implements `toString`
+  toString() : String
+  {
+    ...
+  }
+};
+```
+
+### Fields
+
+Define objects that belong to the class and will be instantiated (ie. use memory) on every instance of the class.
+
+Fields can be initialised in two ways:
+ * `var`: Must be initialised on construction, but can change after.
+ * `let`: Must be initialised on construction, and cannot change later.
```

Comment:
not sure what `var` and `let` have to do with initalisation. I think you mean something different here.

Reply:
IIUC, the idea is to make `let` similar to a `const` field that gets set on construction and cannot change, while `var` is your average field. But I may have expressed it wrong or worse, understood wrong.

Reply:
Both `var` and `let` fields must be initialised at creation time.  We require complete object initialisation because we don't want every field to be an option type with a trivial default and we don't want UB from partial initialisation.

This is a big part of the motivation from splitting `new` (allocate and initialise) from public interfaces for construction.  You want to allow the class to run some code that calculates some default values for things so that you don't need the code that wants a new `Foo` to provide every field for `Foo`, but you *also* don't want to enable partially-constructed `Foo` instances to exist.